developer/hacking-openwisp-python-django.rst

```diff
@@ -0,0 +1,283 @@
+  mathematical expressions, inline plots and other media.
+- Support for interactive data visualization and use of GUI toolkits.
+- Flexible, embeddable interpreters to load into one's own projects
+- Tools for parallel computing
+
+More details including installation and updates can be found on the
+`official webite <https://ipython.org>`_.
+
+As for `ipdb <https://github.com/gotcha/ipdb>`_, it allows to use the
+``ipython`` shell when using the python debugger (known as ``pdb``).
+
+Try to drop this line in a Django project (or an OpenWISP module),
+for example in a ``settings.py`` file:
+
+.. code-block:: python
+
+    import ipdb; ipdb.set_trace()
+
+Now load the Django development server and have fun while learning how
+to do debug Python code!
+
+Django Extensions
+~~~~~~~~~~~~~~~~~
+
+`Django Extensions <https://django-extensions.readthedocs.io/>`_ is a
+collection of extensions for the Django Framework. These include
+management commands, additional database fields, admin extensions and
+much more. We will focus on three of them for now, namely
+``shell_plus``, ``runserver_plus``, ``show_urls``.
+
+Django Extensions can be installed with:
+
+.. code-block:: bash
+
+    pip install django-extensions
+
+`shell_plus
+<django-extensions.readthedocs.io/en/latest/shell_plus.html>`_:
+Django shell with autoloading of the apps database models and subclasses
+of user-defined classes.
+
+`runserver_plus
+<django-extensions.readthedocs.io/en/latest/runserver_plus.html>`_:
+typical runserver with Werkzeug debugger baked in.
+
+``show_urls``: shows the registered URLs of a Django project.
+
+Django Debug Toolbar
+~~~~~~~~~~~~~~~~~~~~
+
+The `Django Debug Toolbar <https://django-debug-toolbar.readthedocs.io/>`_
+is a configurable set of panels which display various debug information
+about the current HTTP request/response and, when clicked, display more
+details about the panel’s content.
+
+It can be installed with:
+
+.. code-block:: bash
+
+    pip install django-debug-toolbar
+
+More information can be found in the `django-debug-toolbar documentation
+<https://django-debug-toolbar.readthedocs.io/en/latest/>`_.
+
+Using these tools in OpenWISP
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These tools can be added to an OpenWISP development environment and
+significantly improve the efficiency and experience while developing.
+Let's do a walkthrough of how to use them in `OpenWISP Controller
+<https://github.com/openwisp/openwisp-controller>`_ as an example. In the
+``tests/`` folder, ``local_settings_example.py`` must be copied and
+renamed to ``local_settings.py`` which we will use for customization.
+This technique can be used in other OpenWISP development environments too.
+
+.. code-block:: bash
+
+  cd tests/
+  cp local_settings_example.py local_settings.py
+
+To start, the `steps <https://github.com/openwisp/openwisp-controller>`_
+for installing OpenWISP Controller must be followed. The commands
+``pipenv run ./manage.py migrate`` and
```

Comment:
Don't we need to do `pipenv install --dev --three --skip-lock` first ?

(Related: in openwisp-controller docs it asks to install via command `pipenv install --dev`, i think `--three` option is required there because most systems take python2.7 as default python version. )

Reply:
@atb00ker Yes, one needs to run at least `pipenv install --dev` first. Once we drop support for Python 2, we can add

    [requires]
    python_version = "3"

to `Pipfile`.